### ISSUE TITLE
Stopping the request handler

### ISSUE THREAD

**Issue by hubertlepicki** [mtrudel/bandit#202]
I have two issues, that are slightly related.

I have noticed that if I leave this line in my Phoenix controllers, the process that handles request never times out, never crashes, it just continues to exit seemingly forever:

```
def index(conn, params) do
  :timer.sleep(:infinity)
  ...
end
```

The process continues to remain alive even if the client browser closes.

The two problems I have with the above behavior are:

1. The request handler does not have a timeout set, and I can't seem to find in Bandit's documentation how to set such timeout. Ideally, I'd like the request handler to be killed after X seconds of not completing it's job - is there a way for me to do it? I only found `:read_timeout` option which is useful but not in this particular case, as I can see from the source of this project that it only gets applied to the time the request body is being read, not to the request handler itself.

2. Is there a way to immediately kill the request handler when the client disconnects (i.e. browser tab closes)?

I have tried adding the following config, but it seems to have no effect either:

```
 config :ui, UI.Endpoint,
   http: [
-    http_1_options: [max_request_line_length: 12288, max_header_length: 12288, compress: true]
+    http_1_options: [max_request_line_length: 12288, max_header_length: 12288, compress: true],
+    thousand_island_options: [read_timeout: 10_000]
   ],
   adapter: Bandit.PhoenixAdapter,
  ...
```

**Comment by hubertlepicki** [comment]
I did some digging and it appears that this is simply not implemented, and I am not sure if that's intentional or not.

@mtrudel I could possibly add the missing functionality if you agree and help me out a bit with directions.

As of 1) the info I would love to hear from you is if it should be handled within Bandit at all, because this can be done on a Plug level. I looked at Cowboy ant it doesn't appear to have such setting either, so my guess is that we don't want it here either.

When it comes to 2) - the ThousandIsland.Handler behavior specifies `handle_close`, which is used in the `Bandit.Websocket.Handler` here https://github.com/mtrudel/bandit/blob/31dbb2f6caed7d0bedfdb566888a30c8b0d125b1/lib/bandit/websocket/handler.ex#L72 but not in (as far as I can tell) Http1 or Http2 handlers.

And even if added, I think the callback would *not* interrupt the above example, with Phoenix controller (or any Plug, really) sleeping to infinity. What I suspect would happen, instead, becaues Handler is a GenServer, is that the messge will be delivered only after the Plug stops working, and it will never interrupt the process that sleeps forever.

I suspect this may be the same issue with Websocket handler.

@mtrudel I am not sure how to best tackle this, and if you agree that 2) is a problem. My feeling is that we should interrupt and shutdown the handler if connection drops. Second best option would be crashing the handler when it tries to write to a closed connection. As it currently is, the Handler continues to execute, and the log lines provided by Plug/Phoenix indicate that response was successfully sent to client (which is not true).

In Cowboy, they silently terminate the handler process if client prematurely closes the connection, which is also not ideal, but at least the process is not left to complete the work that no one waits for completion.

I would very much appreciate your thoughts on the above @mtrudel

**Comment by mtrudel** [comment]
Hi @hubertlepicki , I'll address your original points one by one:

1. Bandit's behaviour with respect to the `:timer.sleep(:infinity)` case is, well not 'by design', definitely acting as expected. To understand why, it helps to think of how Bandit is implemented from a process perspective. Consider HTTP/1 connections; they're implemented as a GenServer (you can see the backing implementation for this in `ThousandIsland.Handler`). As such, the entire life cycle of an HTTP connection (and of any constituent `Plug.call/2` calls) takes places within a single `handle_connection` call. Whatever that process happens to want to do (including sleeping forever) is going to consume that process' entire attention; there's no other place to hang the concern of timing the handler out. If we wanted to have timeout behaviour as you outline, we'd need to have some other process that managed those on behalf of the actual connection process. That's an overhead that's both 1. not terribly idiomatic in the OTP world, and 2. a whole lot of inter-process coordination that slows things down and introduces opportunity for bugs. 

2. With respect to killing a connection as soon as the client disconnects; two things happen when the underlying socket is closed:

* Any calls to access the socket synchronously via `ThousandIsland.Socket` calls will begin returning errors (usually of the form `{:error, :closed}`). These generally get bubbled all the way up to the implementing Plug, but there's a lot of plumbing inside Plug itself that expresses some (not always consistent) logic here as well.

* Once the handler process' execution unrolls out of the current `handle_*` call (if it's in one),  `ThousandIsland.handle_close/2` will be called to give the handler a chance to close things off cleanly.

I think the issue you're having is that you're trying to reconcile expectations of a more fully fleshed out behaviour with the fact that Plug is a pretty bare-bones, one-call, synchronous API. Bandit (and Plug.Cowboy!) do the best they can with this, but it's just by nature an imperfect mapping. 

To summarize, your intuition in your second comment is correct: this behaviour isn't implemented, because there's no straightforward way to do so given the current shape of the Plug API. 

Post Bandit 1.0, I've got some bigger ideas for (breaking) changes to Plug to accommodate things like gRPC that would make some of the things you're asking for more possible, but that's a ways off.

**Comment by mtrudel** [comment]
To the specific case of disconnecting clients, it's expected that Bandit will eventually get around to some sort of network activity, at which point it is expected that the connection shut down. This is the domain of the `read_timeout` parameter you call out, and is really the only place that Bandit (or Cowboy for that matter) has any ability to exhibit timeout semantics.

**Comment by mtrudel** [comment]
All that said, if you can cook up a simple repro I'll happily take a look at improving any corner cases we have!

**Comment by mtrudel** [comment]
Super! I'll dig into your repro and see what we can do. Thanks for the help!

**Comment by mtrudel** [comment]
Once Bandit goes 1.0 I've got some broader plans to rethink the Plug API in order to accommodate things like gRPC that aren't possible given the current Plug semantics. I think this makes sense to look at at that point; there's nothing directly actionable on this until that point I don't think. 

Thanks again for the thoughtful issue!

**Comment by hubertlepicki** [comment]
Okay. Thanks @mtrudel

**Comment by hubertlepicki** [comment]
hey @mtrudel, I am trying to figure out if anything changed here? 

I am thinking about using Bandit again, but this time I'll be making a bunch of analytics requests, which are quite heavy. I do want to interrupt them whenever the client disconnects, to avoid doing unnecessary work. These are some long running SQL queries that no longer have to complete when the user changes filters on the page. On Cowboy thse get cancelled when connection drops, am I right to guess that on Bandit the request handlers still continue to execute when the user browser disconnected?

**Comment by mtrudel** [comment]
You are correct to guess that (at least on HTTP/1). I've been thinking about possible solutions to your original problem ever since you mentioned it months ago and I have a few that I may try out after the latest round of refactors in #297 et al land. They'l likely be opt-in in some fashion, but I'm not exactly sure what they'll look like just yet. I'll make sure to link to them from here if/when they land.
